src/utils.c

```diff
@@ -110,30 +117,221 @@ SEXP colnamesInt(SEXP x, SEXP cols, SEXP check_dups) {
+  int *ansp = INTEGER(ans);
+  for (int i=0; i<=len; i++)
+    ansp[i] = from+i*ord;
+  UNPROTECT(1);
+  return ans;
+}
+
+/* funCols
+ * helper that calls lapply(DT, fun) to produce logical vector
+ */
+SEXP funCols(SEXP x, SEXP fun, SEXP rho) {
+  SEXP lfn = PROTECT(LCONS(sym_lapply, LCONS(x, LCONS(fun, R_NilValue))));
+  SEXP lst = PROTECT(eval(lfn, rho));
+  SEXP log = PROTECT(allocVector(LGLSXP, length(x)));
+  int *logp = LOGICAL(log);
+  for (int i=0; i<length(lst); ++i) {
+    SEXP lsti = VECTOR_ELT(lst, i);
+    if (length(lsti)!=1 || !isLogical(lsti) || LOGICAL(lsti)[0]==NA_LOGICAL)
+      error("When argument for columns selection is a function, it is applied to each column; the output of this function must be a non-missing boolean scalar signalling inclusion/exclusion of the column. However, these conditions were not met at least for: %s", CHAR(STRING_ELT(getAttrib(x, R_NamesSymbol), i)));
+    logp[i] = LOGICAL(lsti)[0];
+  }
+  UNPROTECT(3);
+  return log;
+}
+
+/* exprCols
+ * turns NSE expression into columns int
+ * helper for j and .SDcols
+ * handles v1, "v1", v1:v3, !"v1", -"v1", paste0("v",1:3)
+ * does not evaluate only for `symbol:symbol` expression, or `symbol` when "symbol" is existing column
+ */
+SEXP exprCols(SEXP x, SEXP expr, SEXP mode, /*SEXP with, */SEXP rho) {
+  int protecti=0;
+  if (!isString(mode) || length(mode)!=1 || STRING_ELT(mode, 0)==NA_STRING)
+    error(_("internal error: mode argument has to be character, non-NA of length 1")); // # nocov
+  if (isNull(expr))
+    error(_("columns selection is NULL")); // expr=NULL
+  if (!isNewList(x))
+    error(_("'x' argument must be data.table"));
+  bool mode_j=false, mode_sd=false;
+  if (strcmp(CHAR(STRING_ELT(mode, 0)), "j")==0)
+    mode_j = true;
+  else if (strcmp(CHAR(STRING_ELT(mode, 0)), ".SDcols")==0)
+    mode_sd = true;
+  else
+    error(_("Internal error: invalid 'mode' argument in exprCols function, should have been caught before. please report to data.table issue tracker.")); // # nocov
+  SEXP sym_parenthesis = install("(");
+  SEXP sym_patterns = install("patterns");
+  while (isLanguage(expr) && CAR(expr)==sym_parenthesis)
+    expr = CADR(expr);
+  if (isNull(expr))
+    error(_("columns selection is NULL")); // expr=((NULL))
+  // TODO in colnamesInt: handle inverse selection: !cols_var, !"V2", !V2, !paste0("V",2:3)
+  // but not !V3:V2, we could handle that as well but do we really want?
+  bool inverse = isLanguage(expr) && (CAR(expr)==sym_bang || CAR(expr)==sym_minus);
+  if (inverse)
+    expr = CADR(expr);
+  // non-evaluated case: V3:V2
+  // single symbol V2 may or may not be evaluated, see below
+  if (isLanguage(expr) && CAR(expr)==sym_colon) {  // 3:2, V3:V2, min(V3):min(V2)
+    SEXP lhs = CADR(expr), rhs = CADDR(expr);
+    if (isSymbol(lhs) && isSymbol(rhs)) { // V3:V2
+      if (inverse)
+        error("column range selection by col1:col2 is not supported with combination of negation via '!' or '-'");
```

Comment:
This would break some ```.SDcols``` code as ```!``` is supported.

Reply:
yes, this is still TODO

Reply:
this is covered in recent commit by tests 9999.171 and 9999.172